Remote / Late evaluation of tools

mvdbeek opened the pull request:
The idea is that if we can evaluate tools close to the data we can fetch deferred data there and generate the necessary metadata that the evaluation depends on. This work should be flexible enough to occur within the job script (included in the PR), as a (celery) task that runs on a remote node (maybe with a celery-queue destination option ?) or as a side-pod (will be separate PR(s)).

Tests:
- [x] Unit test for serialising / deserializing JobIO class
  - [x] Make sure ConfiguredFileSources work
- [x] Make sure xml/yml/cwl loading of serialized tool works
- [x] Integration tests that use tool data tables
- [x] Integration tests that use file sources
- [x] Integration test for implicit conversion
- [x] Integration test that exports history (Or maybe not if we merge celery task for history exports ??)
- [x] Unit test for testing FileTracebackException

Development:
- [ ] Do we need to export using export_key ?
- [x] Make  job_wrapper.remote_command_line configurable per destination
- [x] Integrate ToolApp with other app variants
- [x] Enable serializing non-xml tools
- [x] Dataset conversions are probably a hard requirement in practice … can’t have tools randomly fail on a remote_command_line destination based on whether conversion is necessary or not. But should be possible to do, we may just need to serialise ImplicitlyConvertedDatasetAssociation together with the input dataset. Jobs shouldn’t be dispatched before ImplicitlyConvertedDatasetAssociation is ready, so that should be doable. Turns out these don't actually run in the ToolEvaluator -> https://github.com/galaxyproject/galaxy/pull/12991

## How to test the changes?
(Select all options that apply)
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [x] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these contributions under [Galaxy's current license](https://github.com/galaxyproject/galaxy/blob/dev/LICENSE.txt).
- [x] I agree to allow the Galaxy committers to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT). If this condition is an issue, uncheck and just let us know why with an e-mail to galaxy-committers@lists.galaxyproject.org.

mvdbeek committed:
WIP, figure out how to do remote command lines

This kinda works for cat1 at this point, lot's of shortcuts of course.

mvdbeek committed:
Refactor _job_dataset_path_rewriter

mvdbeek committed:
Compose JobWrapper from smaller pieces

mvdbeek committed:
Simplify dataset_path_rewriter property

mvdbeek committed:
Spelling fixes, amend plan

mvdbeek committed:
Add main guard for remote_tool_eval.py

mvdbeek committed:
Fix unit test

mvdbeek committed:
Ugh

mvdbeek committed:
Extend JobIO so it is sufficient to build ComputeEnvironment

mvdbeek committed:
Move to more logical position

mvdbeek committed:
Comment

mvdbeek committed:
Include generated command line in setup

This is all very hacky still

mvdbeek committed:
Rebase: hack for external evaluation of tools that use data tables

mvdbeek commented on lib/galaxy/metadata/remote_tool_eval.py:
```diff
@@ -0,0 +1,77 @@
+        tdtm = ToolDataTableManager.from_dict(json.load(data_tables_json))
+    app = ToolApp(sa_session=import_store.sa_session, tool_app_config=tool_app_config, datatypes_registry=datatypes_registry, object_store=object_store, tool_data_table_manager=tdtm)
+    # TODO: could try to serialize just a minimal tool variant instead of the whole thing ?
+    tool_source = get_tool_source(os.path.join(WORKING_DIRECTORY, 'tool.xml'))
```
Could be a symlink ? Or load from original location ?

mvdbeek committed:
Export model store just once when setting up job

mvdbeek committed:
Read back generate command line in metadata script, attach to job

mvdbeek committed:
Pass builds_file_path and len_file_path in JobIO

jmchilton commented on lib/galaxy/tools/evaluation.py:
```diff
@@ -47,6 +51,48 @@
+Problems:
+Do we want to build a full app ? Probably not.
+Do we want to supply galaxy config file ? Probably not.
+Dataset conversions ... would need to move outside of ToolEvaluator
```
I'm going to take a look at this TODO today - I'll report back if I make any progress on determining if I think it is something I could plausibly help with.

mvdbeek replied:
I have some more work locally, just didn't get around to pushing and updating this here.
I haven't solved / thought about the dataset conversions yet, but I think most of the rest I have figured out.
I'm positive I can wrap this up once we're done with the release process / the obvious bugs in the new release are worked out.

I think in the overall picture it would be good to work out how a deferred dataset would look like in the database and how the UI could look like, any chance you could take a look at this ?

mvdbeek committed:
Fix access to import store objects

mvdbeek committed:
Fix history import/export

mvdbeek committed:
Fix metadata unit tests

mvdbeek committed:
Make dumping tool data table and JobIO conditional on remote_command_line

mvdbeek committed:
Onlt enable remote_command_line when using metadata_strategy: extended

mvdbeek committed:
Export tool_data_path

mvdbeek committed:
Fix JEHA find

mvdbeek committed:
Look up items by id when connecting job_io

mvdbeek committed:
Populate params_dict with short name

mvdbeek committed:
Make remote commandline setup slightly less hacky

mvdbeek committed:
Fix command setup

mvdbeek committed:
Allow serializing YamlToolSource

mvdbeek committed:
Include dataset instances when exporting input dataset collections

Otherwise tests like `test_job_build_for_rerun_list_list` will fail
with:
```
| Dataset Job Standard Error:
|  Traceback (most recent call last):
|  File "metadata/set.py", line 1, in <module>
|  from galaxy_ext.metadata.set_metadata import set_metadata; set_metadata()
|  File "/home/runner/work/galaxy/galaxy/galaxy root/lib/galaxy/metadata/set_metadata.py", line 113, in set_metadata
|  set_metadata_portable()
|  File "/home/runner/work/galaxy/galaxy/galaxy root/lib/galaxy/metadata/set_metadata.py", line 234, in set_metadata_portable
|  import_model_store = store.imported_store_for_metadata('metadata/outputs_new', object_store=object_store)
|  File "/home/runner/work/galaxy/galaxy/galaxy root/lib/galaxy/model/store/__init__.py", line 1583, in imported_store_for_metadata
|  import_model_store.perform_import()
|  File "/home/runner/work/galaxy/galaxy/galaxy root/lib/galaxy/model/store/__init__.py", line 195, in perform_import
|  self._import_collection_instances(object_import_tracker, collections_attrs, history, new_history)
|  File "/home/runner/work/galaxy/galaxy/galaxy root/lib/galaxy/model/store/__init__.py", line 555, in _import_collection_instances
|  dc = import_collection(collection_attrs["collection"])
|  File "/home/runner/work/galaxy/galaxy/galaxy root/lib/galaxy/model/store/__init__.py", line 548, in import_collection
|  materialize_elements(dc)
|  File "/home/runner/work/galaxy/galaxy/galaxy root/lib/galaxy/model/store/__init__.py", line 521, in materialize_elements
|  raise Exception(f"Failed to find HDA with id [{hda_id}] in [{hdas_by_id}]")
|  Exception: Failed to find HDA with id [290] in [{283: <galaxy.model.HistoryDatasetAssociation(283) at 0x7f033e360690>, 284: <galaxy.model.HistoryDatasetAssociation(284) at 0x7f033e3b5290>, 285: <galaxy.model.HistoryDatasetAssociation(285) at 0x7f033e3982d0>, 289: <galaxy.model.HistoryDatasetAssociation(289) at 0x7f033e3cd290>}]
```

mvdbeek committed:
Serialize jobs now

jmchilton commented on lib/galaxy/model/store/__init__.py:
```diff
@@ -1407,9 +1415,11 @@ def add_job_output_dataset_associations(self, job_id, name, dataset_instance):
         self.collections_attrs.append(collection)
         self.included_collections.append(collection)
+        for dataset in collection.collection.dataset_instances:
+            self.add_dataset(dataset, include_files=include_files)
```
I have a variant of this called export_collection in the deferred data branch.

mvdbeek replied:
Awesome, I copied that into https://github.com/galaxyproject/galaxy/pull/12459/commits/2c64e135eaf054d41f6797772286339487dfea7c

mvdbeek committed:
Serialize FileSources

mvdbeek committed:
Don't need to import string after moving version_command

mvdbeek committed:
Converge with deferred data branch

mvdbeek committed:
Fix exporting collections for HDCAs and DatasetCollection instances

mvdbeek committed:
Add typing to ModelExportStore interface

mvdbeek committed:
Fix serialization of refgenie data tables

mvdbeek committed:
More todos and fixmes

mvdbeek committed:
Remove questionable expunge_all

This fixes tests in test/integration/test_pulsar_embedded.py, and it's
not quite clear how that would fix "reverting metadata". My guess is
that there was a race condition, but chances are that there is no race
anymore. If it is actually needed this should go upstream of creating
the JobWrapper object.

mvdbeek committed:
Report traceback if generating commandline fails

mvdbeek committed:
Munge paths for pulsar

Just dropping everything into metadata/outputs_new is not ideal, but
works for now. Much worse is that we break pulsar's path rewriting
(I think we need to at least take into account input path staging
or rewriting).

mvdbeek committed:
Don't export data for implicit collections, but do export input DCEs

mvdbeek committed:
Some more type fixes

mvdbeek committed:
Simplify RemoteToolEvaluator class

I think we can do everything the ToolEvaluator class does eventually,
so we maybe we don't need a special class at all.

mvdbeek committed:
Fix up ToolEvaluatorTestCase

mvdbeek committed:
Move test_version_command into test_evaluation

Since the ToolEvaluator class builds to version command now.

mvdbeek committed:
Fix type hints and abstract classes for ComputeEnvironment

mvdbeek committed:
Fix tasks

mvdbeek committed:
Move remote_tool_eval.py to galaxy.tools

mvdbeek committed:
Resolve circular dependency for ToolEvaluator

by placing ComputeEnvironment and ActionBox under job_execution.

mvdbeek committed:
Fix ToolEvaluator mock setup

mvdbeek committed:
Fix package modules

mvdbeek committed:
Restore JobWrapperTestCase

mvdbeek committed:
Tweak JobIO so we don't need to pass json around if not needed

mvdbeek committed:
Only serialize user-specific file_sources

mvdbeek committed:
Add unit test for JobIO serialization

mvdbeek committed:
Run data manager tests with metadata_strategy: extended

Which currently also means we run with remote_command_line: True.

mvdbeek committed:
Dictifiable typo fix

mvdbeek committed:
Restore ConfiguredFileSources on ToolApp object

mvdbeek committed:
Enable re-import of exported MetdataFile objects

That is needed to reference MetadataFile objects when templating tool
commands and other configfile templates.

mvdbeek committed:
Provide fallback for element_identifiers_extensions_paths_and_metadata_files in sessionless mode

mvdbeek committed:
Use minimal GalaxyDataTestApp instead of MockApp

mvdbeek committed:
Test implicit conversion

mvdbeek committed:
Construct ToolEvaluator only when needed

mvdbeek committed:
Drop plan comment

mvdbeek committed:
Move interactivetool manager setup out of ToolEvaluator

mvdbeek committed:
Run PartialToolEvaluator when setting up job

that should build command line remotely.

We can't quite get away with not running the ToolEvaluator class
completely if a tool has environment variables or interactive tool
entrypoints that have to be templated.

mvdbeek committed:
Load tool directly from JobIO class

mvdbeek committed:
Enable serializing xml/cwl/yaml tool sources via JobIO

mvdbeek committed:
Write externalized commands, append during remote tool evaluation

The externalized commands may for instance set up conda dependency
resolution.

mvdbeek committed:
Test tool derserialization

mvdbeek committed:
Disable remote evaluation for JEHA history export jobs

We'd have the add `JobExportHistoryArchive` to the supported job outputs
in the model export store or move the `JobExportHistoryArchive` creation
from `execute()` ToolAction to the handler and skip creating
`JobExportHistoryArchive` altogether, but we want to perform the
job export in the Celery workers anyway, so I think it's not worth the
effort.
Note that exporting to file sources doesn't use
`JobExportHistoryArchive`, so for those we can still create the command
line remotely (and e.g materialize deferred datasets in a history).

mvdbeek committed:
Make tool_evaluation_strategy configurable on app and destination level

mvdbeek committed:
Replace model hack with Bunch hack

The right fix might be to provide a session on app to avoid both of
these hacks ?

mvdbeek committed:
Add MinimalToolApp container

mvdbeek committed:
Cleanup path handling and add unit test for handling traceback.txt

jmchilton reviewed (approved):
This would have took me twice as long and would have been half as a good, really amazing stuff in here! We will be building cool shit based on this for years - really impressive work.
